net: lib: download_client: Fix address parsing

SeppoTakalo opened the pull request:
Fixed few address parsing issues from download client
* If unsupported protocol was used, for example telnet://example.com parser just assumed that parsing failure as no protocol and defaulted to HTTP. Now returns EPROTONOSUPPORT as documented earlier.
* If no protocol was provided, but beginning of of the host address matched some protocol, that protocol was assumed. For example: download_client_get(&c, "coap.me/index.html", &cfg , 0, 0) uses coap://, because "coap" matches, but it should have used "http://". Now it needs to have "://" in the protocol to match.

Also, converted ztests to native_sim instead of native_posix. Added few more unittests for the cases above.

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-14665/1/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-14665/1/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-fw-nrfconnect-boot | X | | |
| test-fw-nrfconnect-nrf-iot_lwm2m | X | | |
| test-fw-nrfconnect-nrf-iot_mosh | X | | |
| test-fw-nrfconnect-nrf-iot_samples | X | | |
| test-fw-nrfconnect-nrf-iot_serial_lte_modem | X | | |
| test-fw-nrfconnect-nrf-iot_zephyr_lwm2m | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-14665/1/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

SeppoTakalo commented:
Changed the error code in unittests to something that is known by the Arm compiler.
Also added `qemu_cortex_m3` to allowed target, as it uses same compiler than the target HW (nRF91). Removed the `nrf9160_nrf9160_ns` from the list. Makes no sense to run unittests on HW.
